streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals.buffer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Map;
+
+public class InMemoryTimeOrderedKeyValueBufferStore implements StateStore {
```

Comment:
There's a fair number of methods to implement for a `StateStore`, so I created an independent class to take care of all the Store-related bookkeeping. The actual buffer implementation is still in the original class. This seemed to achieve a nice separation of concerns.

It is unlike our other stores, though, so let me know if it's too weird.

Reply:
> This seemed to achieve a nice separation of concerns.

Not sure about this:

With this PR, I don't see why we need to keep the "buffer" and the "state" separated? Logically, the "buffer" was an intermediate thing (incomplete feature) that should be replaced by (or converted into) a `StateStore` with this PR IMHO (ie, merge existing buffer class with this state store class)

Suggestions: rename `TimeOrderedKeyValueBuffer` -> `TimeOrderedKeyValueBufferStore` and let `TimeOrderedKeyValueBufferStore extends StateStore`. Than `InMemoryTimeOrderedKeyValueBufferStore implements `TimeOrderedKeyValueBufferStore`.

Reply:
Ok, I'm also fine with this.